keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,763 @@
+
+### ReplicaSet Status and Deployment Status Changes
+
+We should keep the current counting behavior for `.status.replicas` regardless of the
+PodReplacementPolicy, for backwards compatibility reasons. Current consumers of the Deployment API
+are only expecting non-terminating pods to be present in this field.
+
+To satisfy the requirement for tracking terminating pods, and for implementation purposes,
+we propose a new field `.status.terminatingReplicas` to the ReplicaSet's and Deployment's
+status.
+
+### API
+
+```golang
+// DeploymentPodReplacementPolicy specifies the policy for creating Deployment Pod replacements.
+// Default is a mixed behavior depending on the DeploymentStrategy
+// +enum 
+type DeploymentPodReplacementPolicy string
+const (
+// Terminating policy creates replacement Pods when the old Pods start
+// terminating (has a .metadata.deletionTimestamp). The total number of
+// Deployment Pods can be greater than specified by the Deployment's
+// .spec.replicas and the DeploymentStrategy.
+Terminating DeploymentPodReplacementPolicy = "Terminating"
+// Terminated policy creates replacement Pods only when the old Pods are
+// fully terminated (reach Succeeded or Failed phase). The old Pods are
+// subsequently removed. The total number of the Deployment Pods is
+// limited by the Deployment's .spec.replicas and the DeploymentStrategy.
+Terminated DeploymentPodReplacementPolicy = "Terminated"
+)
+```
+
+```golang
+type DeploymentSpec struct {
+    ...
+    // podReplacementPolicy specifies when to create replacement Pods. 
+	// Possible values are:
+    // - Terminating policy creates replacement Pods when the old Pods start
+	//   terminating (has a .metadata.deletionTimestamp). The total number of
+	//   Deployment Pods can be greater than specified by the Deployment's
+	//   .spec.replicas and the DeploymentStrategy.
+    // - Terminated policy creates replacement Pods only when the old Pods are
+	//   fully terminated (reach Succeeded or Failed phase). The old Pods are
+	//   subsequently removed. The total number of the Deployment Pods is
+	//   limited by the Deployment's .spec.replicas and the DeploymentStrategy.
+    //
+    // The default behavior when the policy is not specified depends on the DeploymentStrategy:
+	// - Recreate strategy uses Terminated behavior when recreating the deployment,
+	//   but uses Terminating when scaling the deployment.
+	// - RollingUpdate strategy uses Terminating behavior for both rolling out and
+	//   scaling the deployments.
+	//
+	// This is an alpha field. Enable DeploymentPodReplacementPolicy to be able to
+	// use this field.
+    // +optional
+    PodReplacementPolicy *DeploymentPodReplacementPolicy
+    ...
+}
+```
+
+```golang
+type ReplicaSetStatus struct {
+    ...
+	// The number of terminating replicas for this replica set.
+	//
+	// This is an alpha field. Enable DeploymentPodReplacementPolicy to be able to
+	// use this field.
+	// +optional
+	TerminatingReplicas int32
```

Comment:
Curious about relationships between counters here too. I see "non-terminated" used in comments, I assume that is different than what "terminatingReplicas" will count but it would be best if we state it as clearly as possible.

```go
	// Total number of non-terminated pods targeted by this deployment (their labels match the selector).
	// +optional
	Replicas int32

	// Total number of non-terminated pods targeted by this deployment that have the desired template spec.
	// +optional
	UpdatedReplicas int32

	// Total number of ready pods targeted by this deployment.
	// +optional
	ReadyReplicas int32

	// Total number of available pods (ready for at least minReadySeconds) targeted by this deployment.
	// +optional
	AvailableReplicas int32

	// Total number of unavailable pods targeted by this deployment. This is the total number of
	// pods that are still required for the deployment to have 100% available capacity. They may
	// either be pods that are running but not yet available or pods that still have not been created.
	// +optional
	UnavailableReplicas int32
```

Reply:
> This looks right, but just to confirm, +optional indicates that this will be an omitempty field?

Yes. I have added the tags for clarity.

> Curious about relationships between counters here. I see "non-terminated" used in comments, I assume that is different than what "terminatingReplicas" will count but it would be best if we state it as clearly as possible.

I have also added the other relevant fields to the KEP.

I have replaced `non-terminated` with `non-terminating` as I think it describes that better. And it should include pods that do not have a `.metadata.deletionTimestamp` field set. 

Terminating pods have  a `.metadata.deletionTimestamp` field set.

I have tried to make the fields clearer. But having `non-terminating (no .metadata.deletionTimestamp)` everywhere seems too verbose, so I have added the full explanation only to the `TerminatingReplicas` field

> Also, do any of the Replicas, UpdatedReplicas, ... counts include the TerminatingReplicas count?
> Which counters (if any) also count terminatingReplicas?

No, they do not. And they cannot - to be backwards compatible.

> Is there any relationship between the counts (e.g. does ReadyReplicas + TerminatingReplicas = Replicas?)

There is not. Basically the amount of terminating replicas is unlimited. For example if you have `Terminating` policy and do multiple rollouts in a row and have a `TerminationGracePeriodSeconds`.